goworker.go

```diff
@@ -87,12 +90,71 @@ func Init() error {
 // while they wait for an available connection. Expect this
 // API to change drastically.
 func GetConn() (*RedisConn, error) {
+	if workerSettings.RedisSettings.MasterName == "" {
+		return getConn()
+	} else {
+		return getConnSentinel()
+	}
+}
+
+func getConn() (*RedisConn, error) {
 	resource, err := pool.Get(ctx)
 
 	if err != nil {
 		return nil, err
 	}
 	return resource.(*RedisConn), nil
+
+}
+func getConnSentinel() (*RedisConn, error) {
+	deadConnection := errors.New("Dead connection")
+	slaveConnection := errors.New("Stale connection (to slave, not master)")
+	try := func() (*RedisConn, error) {
+		resource, err := pool.Get(ctx)
+
+		if err != nil {
+			return nil, err
+		}
+		conn := resource.(*RedisConn)
+
+		// close the connection and remove it from the pool so that new
+		// connections get created
+
+		// if the instance does not ping back
+		_, err = conn.Do("ping")
+		if err != nil {
+			conn.Close()
+			PutConn(nil)
+			return nil, deadConnection
+		}
+
+		// or if the instance is not a master anymore
+		role, err := redis.String(conn.Do("role"))
```

Comment:
I'm not sure about the behavior or `role` on a stand-alone Redis but even if it behaved the same, it wouldn't help us. With Sentinel we want to cycle through the pool of connections, check them, throw out the dead ones or those connected to slave, so that eventually we get to ask Sentinel for the current master and try that one. Without Sentinel, we can do the role check but a dead connection should be returned immediately to be handled by the caller.
Role check could be done in either case, but I can't think of a situation when it would be useful to try multiple connections from the pool for a stand-alone if one of them doesn't ping back. On the other hand, I don't know that much about Redis.